metrics_batcher.go

```diff
@@ -89,3 +124,65 @@ func (mb *metricsBatcher) finalError() error {
 	}
 	return fmt.Errorf("[%s]", strings.Join(errMsgs, "; "))
 }
+
+// sendReq sends create time series requests to Stackdriver,
+// and returns the count of dropped time series and error.
+func sendReq(ctx context.Context, c *monitoring.MetricClient, req *monitoringpb.CreateTimeSeriesRequest) (int, error) {
+	err := createTimeSeries(ctx, c, req)
+	if err != nil {
+		return len(req.TimeSeries), err
+	}
+	return 0, nil
+}
+
+type worker struct {
+	ctx context.Context
+	mc  *monitoring.MetricClient
+
+	resp *response
+
+	quit     chan bool
+	respChan chan *response
+	reqChan  chan *monitoringpb.CreateTimeSeriesRequest
+}
+
+func newWorker(ctx context.Context, mc *monitoring.MetricClient) *worker {
+	return &worker{
+		ctx:      ctx,
+		mc:       mc,
+		resp:     &response{},
+		reqChan:  make(chan *monitoringpb.CreateTimeSeriesRequest),
```

Comment:
why not only one channel shared by the workers? See https://gobyexample.com/worker-pools

Same for responses.

Reply:
Updated to use shared chans.

Reply:
Took back what I said - using a shared channel for requests is not applicable here, because we don't know how many times each worker will be called (as compared to https://gobyexample.com/worker-pools it has a fixed number of execution). For response chan it's ok to use shared chan since each worker will return exactly one response.

Reply:
Not sure I understand. I expect that every worker will pick one item from the reqsChan when idle, so no need to explicitly assign work to one worker.

Also they will know if no more work left if the reqsChan gets closed.

Reply:
> I expect that every worker will pick one item from the reqsChan when idle, so no need to explicitly assign work to one worker.

The problem is if we have only one shared channel, when doing
```go
	for {
		select {
		case req := <-sharedReqsChan:
			w.recordDroppedTimeseries(sendReq(w.ctx, w.mc, req))
		case <-w.quit:
			respChan <- w.resp
			return
		}
	}
```
Workers will always try to get something from `sharedReqsChan` even if it only returns nil. The only way I can think of to make it work is:
```go
	for {
		select {
		case req := <-sharedReqsChan:
                        if req != nil {
                          w.recordDroppedTimeseries(sendReq(w.ctx, w.mc, req))
                        }
		case <-w.quit:
			respChan <- w.resp
			return
		}
	}
```
Which is kind of a busy-waiting. Do you have a better suggestion?

Reply:
Updated in https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/pull/213/commits/c8c87168d8d57e3d2c941740f1d28f74297378d2.